test/unit/org/apache/cassandra/transport/ClientResourceLimitsTest.java

```diff
@@ -137,6 +143,23 @@ private SimpleClient client(boolean throwOnOverload, int largeMessageThreshold)
         }
     }
 
+    private void emulateInFlightConcurrentMessage(long length)
+    {
+        ClientResourceLimits.Allocator allocator = ClientResourceLimits.getAllocatorForEndpoint(FBUtilities.getJustLocalAddress());
+        ClientResourceLimits.ResourceProvider resourceProvider = new ClientResourceLimits.ResourceProvider.Default(allocator);
+        resourceProvider.globalLimit().allocate(length);
+        resourceProvider.endpointLimit().allocate(length);
+        emulatedUsedCapacity += length;
+    }
+
+    private void releaseEmulatedCapacity(long length)
+    {
+        ClientResourceLimits.Allocator allocator = ClientResourceLimits.getAllocatorForEndpoint(FBUtilities.getJustLocalAddress());
+        ClientResourceLimits.ResourceProvider resourceProvider = new ClientResourceLimits.ResourceProvider.Default(allocator);
+        resourceProvider.globalLimit().release(length);
+        resourceProvider.endpointLimit().release(length);
+    }
```

Comment:
nit: Noting that this is duplicated in `RateLimitingTest`. Might be something we could move to a shared abstract class, along with query options and client creation, but not a big deal if you don't think it's worth it. `MessageSizeLimitTest` and `AuthMessageSizeLimitTest` share some logic as well. Perhaps `abstract MessageLimitTest` or `abstract NativeProtocolLimitTest`?

Reply:
agree, I have extracted some common logic from these 4 classes